images/converter/default.go

```diff
@@ -60,41 +60,93 @@ type ConvertHooks struct {
 
 // IndexConvertFuncWithHook is the convert func used by Convert with hook functions support.
 func IndexConvertFuncWithHook(layerConvertFunc ConvertFunc, docker2oci bool, platformMC platforms.MatchComparer, hooks ConvertHooks) ConvertFunc {
-	c := &defaultConverter{
+	return newDefaultWithOpts(convertOpts{
 		layerConvertFunc: layerConvertFunc,
 		docker2oci:       docker2oci,
 		platformMC:       platformMC,
-		diffIDMap:        make(map[digest.Digest]digest.Digest),
 		hooks:            hooks,
+	}).Convert
+}
+
+func NewDefault(opts ...Opt) (*DefaultConverter, error) {
+	var copts convertOpts
+	for _, o := range opts {
+		if err := o(&copts); err != nil {
+			return nil, err
+		}
+	}
+
+	return newDefaultWithOpts(copts), nil
+}
+
+func newDefaultWithOpts(copts convertOpts) *DefaultConverter {
+	if copts.platformMC == nil {
+		copts.platformMC = platforms.All
+	}
+
+	return &DefaultConverter{
+		convertOpts: copts,
+	}
+}
+
+type DefaultConverter struct {
+	convertOpts
+}
+
+type diffIDMap struct {
+	diffIDs map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
+	mutex   sync.RWMutex
+}
+
+func makeRefKey(desc ocispec.Descriptor) (string, error) {
+	key, err := json.Marshal(desc)
+	if err != nil {
+		return "", err
 	}
-	return c.convert
+	return string(key), nil
 }
 
-type defaultConverter struct {
-	layerConvertFunc ConvertFunc
-	docker2oci       bool
-	platformMC       platforms.MatchComparer
-	diffIDMap        map[digest.Digest]digest.Digest // key: old diffID, value: new diffID
-	diffIDMapMu      sync.RWMutex
-	hooks            ConvertHooks
+// singleflightDo makes sure that only one conversion execution is
+// in-flight for a given descriptor at a time.
+func (c *DefaultConverter) singleflightDo(ctx context.Context, cs content.Store, desc ocispec.Descriptor, dm *diffIDMap, convert convertFuncWithDiffMap) (*ocispec.Descriptor, error) {
+	if c.sg == nil {
+		return convert(ctx, cs, desc, dm)
+	}
+
+	key, err := makeRefKey(desc)
+	if err != nil {
+		return nil, err
+	}
+
+	newDesc, err, _ := c.sg.Do(key, func() (interface{}, error) {
+		return convert(ctx, cs, desc, dm)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newDesc.(*ocispec.Descriptor), nil
 }
 
 // convert dispatches desc.MediaType and calls c.convert{Layer,Manifest,Index,Config}.
 //
 // Also converts media type if c.docker2oci is set.
-func (c *defaultConverter) convert(ctx context.Context, cs content.Store, desc ocispec.Descriptor) (*ocispec.Descriptor, error) {
+func (c *DefaultConverter) Convert(ctx context.Context, cs content.Store, desc ocispec.Descriptor) (*ocispec.Descriptor, error) {
 	var (
 		newDesc *ocispec.Descriptor
 		err     error
 	)
+	dm := &diffIDMap{
+		diffIDs: make(map[digest.Digest]digest.Digest),
+	}
```

Comment:
This doesn't seems to work and produces a broken config blob when a layer converter produces a layer having different diffID from the original one.

e.g. The following test fails

```golang
package testconverter

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"testing"

	"github.com/containerd/containerd/content"
	"github.com/containerd/containerd/errdefs"
	"github.com/containerd/containerd/images"
	"github.com/containerd/containerd/images/converter"
	"github.com/containerd/containerd/platforms"
	"github.com/containerd/stargz-snapshotter/util/testutil"
	"github.com/opencontainers/go-digest"
	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
)

func TestLayerConvertFunc(t *testing.T) {
	ctx := context.Background()
	desc, cs, err := testutil.EnsureHello(ctx)
	if err != nil {
		t.Fatal(err)
	}

	lcf := LayerConvertFunc() // Produce a layer having different diffID than the original
	docker2oci := true
	platformMC := platforms.DefaultStrict()
	cf := converter.DefaultIndexConvertFunc(lcf, docker2oci, platformMC)

	newDesc, err := cf(ctx, cs, *desc)
	if err != nil {
		t.Fatal(err)
	}

	var wantDiffID string
	var gotDiffID string
	handler := func(hCtx context.Context, hDesc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
		if !images.IsConfigType(hDesc.MediaType) && !images.IsLayerType(hDesc.MediaType) {
			return nil, nil
		}
		if images.IsConfigType(hDesc.MediaType) {
			ra, err := cs.ReaderAt(ctx, hDesc)
			if err != nil {
				return nil, err
			}
			defer ra.Close()
			diffID := digest.Canonical.Digester()
			buf := new(bytes.Buffer)
			if _, err := io.Copy(diffID.Hash(), io.TeeReader(io.NewSectionReader(ra, 0, ra.Size()), buf)); err != nil {
				return nil, err
			}
			dt := buf.Bytes()
			var image ocispec.Image
			if err := json.Unmarshal(dt, &image); err != nil {
				return nil, err
			}
			gotDiffID = image.RootFS.DiffIDs[0].String()
		} else {
			wantDiffID = hDesc.Digest.String()
		}
		return nil, nil
	}
	handlers := images.Handlers(
		images.ChildrenHandler(cs),
		images.HandlerFunc(handler),
	)
	if err := images.Walk(ctx, handlers, *newDesc); err != nil {
		t.Fatal(err)
	}
	if gotDiffID == "" || wantDiffID == "" {
		t.Fatalf("lack diff ids")
	}
	if gotDiffID != wantDiffID {
		t.Fatalf("wanted %q; got %q", wantDiffID, gotDiffID)
	}
}

func LayerConvertFunc() converter.ConvertFunc {
	return func(ctx context.Context, cs content.Store, desc ocispec.Descriptor) (*ocispec.Descriptor, error) {
		if !images.IsLayerType(desc.MediaType) {
			// No conversion. No need to return an error here.
			return nil, nil
		}
		ref := fmt.Sprintf("convert-test-from-%s", desc.Digest)
		w, err := content.OpenWriter(ctx, cs, content.WithRef(ref))
		if err != nil {
			return nil, err
		}
		defer w.Close()

		// Reset the writing position
		// Old writer possibly remains without aborted
		// (e.g. conversion interrupted by a signal)
		if err := w.Truncate(0); err != nil {
			return nil, err
		}
		testlayer := testutil.BuildTar([]testutil.TarEntry{testutil.File("test", "testvalue")})
		n, err := io.Copy(w, testlayer)
		if err != nil {
			return nil, err
		}
		if err = w.Commit(ctx, n, ""); err != nil && !errdefs.IsAlreadyExists(err) {
			return nil, err
		}
		if err := w.Close(); err != nil {
			return nil, err
		}
		newDesc := desc
		newDesc.MediaType = ocispec.MediaTypeImageLayer
		newDesc.Digest = w.Digest()
		newDesc.Size = n
		newDesc.Annotations = nil
		return &newDesc, nil
	}
}

```

Reply:
👍 Thanks for the testing! Fixed, will add a test case later.

Reply:
@ktock Updated the test case for diff id checks in the third commit, PTAL.